lib/useOnyx.ts

```diff
@@ -0,0 +1,200 @@
+    const isFirstRenderRef = useRef(true);
+    const fetchStatusRef = useRef<FetchStatus>('loading');
+
+    // eslint-disable-next-line rulesdir/prefer-early-return
+    useEffect(() => {
+        /**
+         * This condition will ensure we can only handle dynamic collection member keys.
+         */
+        if (previousKey !== key && !(isCollectionMemberKey(previousKey) && isCollectionMemberKey(key))) {
+            throw new Error(
+                `'${previousKey}' key can't be changed to '${key}'. useOnyx() only supports dynamic keys if they are both collection member keys e.g. from 'collection_id1' to 'collection_id2'.`,
+            );
+        }
+    }, [previousKey, key]);
+
+    /**
+     * According to React docs, `getSnapshot` is a function that returns a snapshot of the data in the store that’s needed by the component.
+     * **While the store has not changed, repeated calls to getSnapshot must return the same value.**
+     * If the store changes and the returned value is different (as compared by Object.is), React re-renders the component.
+     */
+    const getSnapshot = useCallback(() => {
+        /**
+         * Case 1 - We have a normal key without selector
+         *
+         * We just return the data from the Onyx cache.
+         */
+        if (!Onyx.isCollectionKey(key) && !options?.selector) {
+            return getCachedValue(key) as TReturnData;
+        }
+
+        /**
+         * Case 2 - We have a normal key with selector
+         *
+         * Since selected data is not directly stored in the cache, we need to generate it with `getCachedValue`
+         * and deep compare with our previous internal data.
+         *
+         * If they are not equal, we update the internal data and return it.
+         *
+         * If they are equal, we just return the previous internal data.
+         */
+        if (!Onyx.isCollectionKey(key) && options?.selector) {
+            const newData = getCachedValue(key, options.selector);
+            if (!deepEqual(previousDataRef.current, newData)) {
+                previousDataRef.current = newData as TReturnData;
+            }
+
+            return previousDataRef.current as TReturnData;
+        }
+
+        /**
+         * Case 3 - We have a collection key with/without selector
+         *
+         * Since both collection objects and selected data are not directly stored in the cache, we need to generate them with `getCachedValue`
+         * and deep compare with our previous internal data.
+         *
+         * If they are not equal, we update the internal data and return it.
+         *
+         * If they are equal, we just return the previous internal data.
+         */
+        const newData = getCachedValue(key, options?.selector);
+        if (!deepEqual(previousDataRef.current, newData)) {
+            previousDataRef.current = newData as TReturnData;
+        }
+
+        return previousDataRef.current as TReturnData;
```

Comment:
This is the same code as above in Case 2, right?

Reply:
Is almost the same, in Case 2 we have a condition. I kept them separated to explain better the possible cases.